content/posts/2024-11-05-flang-new.md

```diff
@@ -0,0 +1,527 @@
+
+This lead languages such as Swift, Rust and Julia to create their own IRs
+which include information useful for their own optimisations.
+After that IR has been optimised it is lowered into LLVM IR and goes through the
+normal pipeline.
+
+In doing so, they have to implement a lot of common things repeatedly. This is
+where MLIR comes in.
+
+> The MLIR project aims to directly tackle these programming language design and
+> implementation challenges—by making it very cheap to define and introduce new
+> abstraction levels, and provide “in the box” infrastructure to solve common
+> compiler engineering problems.
+- ["MLIR: A Compiler Infrastructure for the End of Moore’s Law"](https://arxiv.org/abs/2002.11054)
+  (Chris Lattner et al., 2020)
+
+# Flang and MLIR
+
+The same year MLIR debuted, Eric Schweitz (NVIDIA) gave a talk at the later US
+LLVM Developers' meeting titled
+["An MLIR Dialect for High-Level Optimization of Fortran”](https://www.youtube.com/watch?v=ff3ngdvUang).
+
+Remember the `fir-dev` branch? The `fir` is this.
+[Fortran Intermediate Representation](https://flang.llvm.org/docs/FIRLangRef.html),
+better known to developers as "FIR".
+
+FIR did the same thing that Rust, Swift and Julia did but this time, Flang had
+the benefit of basing FIR on MLIR.
+
+They did not stop there, later adding
+["High Level Fortran Intermediate Representation"](https://flang.llvm.org/docs/HighLevelFIR.html)
+(HLFIR) which works at a higher level than FIR. A big target of HLFIR 
+was array optimisations, which were more complex to handle using just FIR.
+
+## IRs All the Way Down
+
+We start with Fortran source code. This function assigns array `b` to array `a`.
+
+```fortran
+subroutine foo(a, b)
+  real :: a(:), b(:)
+  a = b
+end subroutine
+```
+
+(if you want to modify this example, you can do so on
+[Compiler Explorer](https://godbolt.org/z/851qrcj6a))
+
+It is tempting to think of the IRs in a "stack" where each one is converted
+into the next. However, MLIR allows multiple "dialects" of MLIR to exist in the
+same file.
+
+As you can see in this first step, Flang produces a file that uses the HLFIR
+dialect where it can, and the FIR dialect for the rest.
+
+```mlir
+module attributes {<...>} {
+  func.func @_QPfoo(%arg0: !fir.box<!fir.array<?xf32>> {fir.bindc_name = "a"}, %arg1: !fir.box<!fir.array<?xf32>> {fir.bindc_name = "b"}) {
+    %0 = fir.dummy_scope : !fir.dscope
+    %1:2 = hlfir.declare %arg0 dummy_scope %0 {uniq_name = "_QFfooEa"} : (!fir.box<!fir.array<?xf32>>, !fir.dscope) -> (!fir.box<!fir.array<?xf32>>, !fir.box<!fir.array<?xf32>>)
+    %2:2 = hlfir.declare %arg1 dummy_scope %0 {uniq_name = "_QFfooEb"} : (!fir.box<!fir.array<?xf32>>, !fir.dscope) -> (!fir.box<!fir.array<?xf32>>, !fir.box<!fir.array<?xf32>>)
+    hlfir.assign %2#0 to %1#0 : !fir.box<!fir.array<?xf32>>, !fir.box<!fir.array<?xf32>>
+    return
+  }
+}
+```
+For example, the parameters are declared with `hlfir.declare` but their type is
```

Comment:
If you want to use Fortran terminology here, `parameters` must be replaced by `dummy arguments`.

Reply:
I've used the term and linked to FortranLang's explanation of it, to save space.